### PR TITLE
Cut 2.20.0-rc.0

### PR DESCRIPTION
**Pull request by beorn7**
There are a bunch of bug fixes in flight that I would like to see in
if possible. Thus, I'm postponing the RC by one day, but I wanted to
get some early reviews of the CHANGELOG already.

### CODE REVIEW

**Comment by beorn7** [comment]
Adding some reviewers that I felt were particularly involved in this release.

Also, @RichiH , FYI so that you know what's coming.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,35 @@
...
-
-* [ENHANCEMENT] TSDB: WAL compression is enabled by default.
+## 2.20.0-rc.0 / 2020-07-16
+
+* [CHANGE] promtool: Changed rule numbering from 0-based to 1-based when reporting rule errors. #7495
+* [CHANGE] Remote read: Added `prometheus_remote_storage_read_queries_total` counter and `prometheus_remote_storage_read_request_duration_seconds` histogram, removed `prometheus_remote_storage_remote_read_queries_total` counter.
+* [CHANGE] TSDB: WAL compression is enabled by default. #7410
+* [FEATURE] PromQL: Added `group()` aggregator. #7480
+* [FEATURE] SD: Added Docker Swarm SD. #7420
+* [FEATURE] SD: Added Digital Ocean SD. #7407
```
Comment (roidelapluie) [inline CHANGELOG.md:8]:
```suggestion
* [FEATURE] SD: Added DigitalOcean SD. #7407
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,35 @@
...
+* [FEATURE] SD: Added Openstack config option to query alternative endpoints. #7494
+* [ENHANCEMENT] Configuration: Exit early on invalid config file and signal it with exit code 2. #7399
+* [ENHANCEMENT] Documentation: Added Docker Swarm configuration example. #7542
```
Comment (roidelapluie) [inline CHANGELOG.md:11]: That is part of the new SD feature

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,7 +1,35 @@
...
+* [ENHANCEMENT] SD: Use 10m timeout for Consul watches. #7423
+* [ENHANCEMENT] SD: Added AMI meta label for EC2 SD. #7386
+* [ENHANCEMENT] TSDB: Log the start of creating a checkpoint. #7581
+* [ENHANCEMENT] TSDB: Increment WAL corruption metric also on WAL corruption during checkpointing. #7491
+* [ENHANCEMENT] TSDB: Improved query performance for high-cardinality labels. #7448
+* [ENHANCEMENT] UI: Display dates as well as timestamps in status page. #7544
+* [ENHANCEMENT] UI: Improved scrolling when following hash-fragment links. #7456
+* [ENHANCEMENT] UI: React UI renders numbers in alerts in a more human-readable way. #7426
+* [BUGFIX] API: Fixed error status code in the query API. #7435 
+* [BUGFIX] PromQL: Fixed `avg_over_time` for NaN and float64 overflows. #7346
```
Comment (roidelapluie) [inline CHANGELOG.md:26]:
```suggestion
* [BUGFIX] PromQL: Fixed `avg` and `avg_over_time` for NaN, Inf, and float64 overflows. #7346
```

File: CHANGELOG.md (hunk 2 above)
Comment (beorn7) [reply CHANGELOG.md:11]: Of course, will remove.

**Comment by beorn7** [comment]
Thanks @roidelapluie . All suggestion have been applied, but GH has a problem right now, so they don't show up yet.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,7 +1,34 @@
...
+* [FEATURE] SD: Added Openstack config option to query alternative endpoints. #7494
+* [ENHANCEMENT] Configuration: Exit early on invalid config file and signal it with exit code 2. #7399
+* [ENHANCEMENT] PromQL: `without` is now a valid metric identifier. #7533
+* [ENHANCEMENT] PromQL: Optimized regex label matching for literals within the pattern or as prefix/suffix. #7453 #7503
 * [ENHANCEMENT] promtool: Added time range parameters for labels API in promtool. #7463
+* [ENHANCEMENT] Remote write: Include samples waiting in channel in pending samples metric. Log number of dropped samples on hard shutdown. #7335
+* [ENHANCEMENT] SD: Reduce timeouts for Openstack SD. #7507
+* [ENHANCEMENT] SD: Use 10m timeout for Consul watches. #7423
+* [ENHANCEMENT] SD: Added AMI meta label for EC2 SD. #7386
+* [ENHANCEMENT] TSDB: Log the start of creating a checkpoint. #7581
```
Comment (brian-brazil) [inline CHANGELOG.md:18]: Is this worth mentioning? Minor log changes aren't usually mentioned.

**Comment by roidelapluie** [comment]
That github UI not up to date is annoying.

File: CHANGELOG.md (hunk 1 above)
Comment (roidelapluie) [reply CHANGELOG.md:8]: That is not fixed

File: CHANGELOG.md (hunk 4 above)
Comment (beorn7) [reply CHANGELOG.md:18]: I thought it's useful to know for those that need this log line. But I don't feel strongly. I'll remove this line unless someone speaks up in favor of keeping it very soon.

File: CHANGELOG.md (hunk 1 above)
Comment (beorn7) [reply CHANGELOG.md:8]: Oops. Will fix in next push.

File: CHANGELOG.md (hunk 4 above)
Comment (beorn7) [reply CHANGELOG.md]: Removed!

**Comment by beorn7** [comment]
Added #7342 and #7323, found by @brian-brazil (thanks for the catch, they were so close to the 2.19 release that I missed them in the `git log` stream).

Also added recently merged #7573. 

I'd really like to get #7562 in for this release. It seems good to go but needs final approval by a maintainer.

**Comment by brian-brazil** [comment]
:+1:

**Comment by beorn7** [comment]
OK, this should be ready now. I'll wait for at least one final :+1: .

**Comment by brian-brazil** [comment]
:+1:

**Comment by beorn7** [comment]
Just to be super sure, I'll wait for all tests to pass.

**Comment by codesome** [comment]
Memory RSS is less for this release, I wonder where that is coming from. 
Also, the query timings look weird, as if some part of inner_eval time was shifted to prepare_time in 2.20, but overall the query times seem to be less if you add all of them up (could it be any of #7453 #7503 #7448?).

**Comment by roidelapluie** [comment]
The CPU usage is also less important.

**Comment by codesome** [comment]
The low watermark is also very low compared to 2.19. I think that is https://github.com/prometheus/prometheus/pull/7562 for the reduction in CPU and the number of watermarks (hence commits to TSDB)?

**Comment by codesome** [comment]
That can explain memory usage reduction too

**Comment by roidelapluie** [comment]
I have executed prombench on that #7562 PR and I did not notice that.

**Comment by beorn7** [comment]
Perhaps somehow, you benchmarked the wrong diff? I mean, the watermark increase _has_ to approximately half with your change, and that's what we are seeing now.

**Comment by roidelapluie** [comment]
That was seen on that prombench too.

**Comment by roidelapluie** [comment]
I mean: I did notice the watermark change but did not notice the CPU/Memory changes.

**Comment by beorn7** [comment]
OK, cool.

Then no red flags so far, on the contrary.

**Comment by roidelapluie** [comment]
@beorn7 Before cancelling the benchmark, please take and save a 5 minute CPU profile on both instances on high load, so we have a chance to understand what is going on, WDYT?

**Comment by roidelapluie** [comment]
high load means once the part of the cycle where all the targets expose many data

**Comment by roidelapluie** [comment]
As we are running for 3 days I have ran pprof ; the winner is.... :partying_face:  @pracucci :tada: with #7390 & #7413

[pprof.zip](https://github.com/prometheus/prometheus/files/4944576/pprof.zip)

**Comment by roidelapluie** [comment]
m1
m2

Thanks @pracucci

**Comment by beorn7** [comment]
@roidelapluie I assume you have taken the pprof profiles you wanted. I'll just cancel the benchmark now. Will then cut rc.1 with the (few) changes so far. Just to play safe…

**Comment by bwplotka** [comment]
:tada: Awesome work all!

**Comment by roidelapluie** [comment]
Yes we have an explanation about the CPU difference. The fact that it is on the read path also explains why the difference can also be seen when the CPU usage is lower.

**Pull request merged**